[feat] Add reusable ChatInput audio components

sfc-gh-nbellante opened the pull request:
## Describe your changes

Added new chat input components for the Streamlit frontend:

1. `ChatComposer` - A versatile chat input component that supports:
    - Text input with keyboard shortcuts (Enter to send, Shift+Enter for new line)
    - File attachments via a file picker
    - Audio recording capabilities when enabled
2. `ChatAudioRecorder` - A component for recording audio messages that:
    - Displays a waveform visualization during recording
    - Provides approve/cancel controls for the recording
    - Handles audio format conversion for different browser implementations

These components are designed to be used together to create a modern chat interface with multimedia support.

## Testing Plan

- Unit Tests: Added comprehensive test suites for both components:
    - `ChatAudioRecorder.test.tsx` - Tests recording lifecycle, error handling, and UI states
    - `ChatComposer.test.tsx` - Tests input handling, file selection, and audio recording integration
- The tests cover key functionality including:
    - Recording start/stop/cancel flows
    - Error handling during recording
    - Accessibility features
    - State management during recording
    - File format handling
- Manual testing should include:
    - Testing in different browsers to verify audio recording compatibility
    - Testing with screen readers to verify accessibility
    - Testing the responsive behavior of the components

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

sfc-gh-nbellante committed:
Add reusable ChatInput audio components

sfc-gh-nbellante commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12834?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#12834** <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12834?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12834?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#12833** <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12833?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>: 1 other dependent PR ([#12835](https://github.com/streamlit/streamlit/pull/12835) <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12835?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>)
* **#12832** <a href="https://app.graphite.dev/github/pr/streamlit/streamlit/12832?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `develop`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

sfc-gh-nbellante committed:
Fix linting error in ChatComposer test

- Extract MockChatAudioRecorder as a proper named component

sfc-gh-nbellante committed:
Fix ChatComposer to provide required containerRef

Add waveformContainerRef to ChatComposer and pass it to
useWaveformController to satisfy the required containerRef parameter.

Changes:
- Create waveformContainerRef using useRef
- Pass containerRef to useWaveformController

This fixes the TypeScript error after making containerRef required
in the useWaveformController interface.

sfc-gh-nbellante commented on frontend/lib/src/components/ChatInput/ChatAudioRecorder.tsx:
```diff
@@ -0,0 +1,225 @@
+  controller: WaveformController
+  isRecording: boolean
+  onStart: () => void | Promise<void>
+  onApprove: (payload: { blob: Blob; meta: AudioMeta }) => void | Promise<void>
```
**suggestion:** Make this always return Promise<void> for simplicity

sfc-gh-nbellante replied:
Fixed. The `onStart` callback now has an explicit `Promise<void>` return type for consistency.

sfc-gh-nbellante commented on frontend/lib/src/components/ChatInput/ChatAudioRecorder.tsx:
```diff
@@ -0,0 +1,225 @@
+  }
+}
+
+const DEFAULT_STRINGS: Required<
```
**suggestion:** this seems like over-engineering. Why do the strings need to be replacable?

sfc-gh-nbellante replied:
Fixed. Removed the customizable strings pattern from ChatAudioRecorder as it was over-engineering. Labels are now hard-coded directly in the component.

sfc-gh-nbellante commented on frontend/lib/src/components/ChatInput/ChatAudioRecorder.tsx:
```diff
@@ -0,0 +1,225 @@
+
+    const startRecording = async (): Promise<void> => {
+      // Wait for the ref to be committed to the DOM
+      await new Promise(resolve =>
```
**question:** Is this necessary? What happens if you don't do this?

sfc-gh-nbellante commented on frontend/lib/src/components/ChatInput/ChatAudioRecorder.tsx:
```diff
@@ -0,0 +1,225 @@
+      }
+    }
+
+    startRecording().catch(_error => {
```
**question:** What sort of promise rejections can happen in startRecording? In general, it's a bad practice to silently swallow errors like this.

sfc-gh-nbellante replied:
Fixed. Errors are now logged using the `loglevel` library (imported as `LOG`) for debugging and monitoring. The error is still handled by calling `onCancel()` to notify the parent component, but it's no longer silently swallowed—it's explicitly logged for observability.

sfc-gh-nbellante commented on frontend/lib/src/components/ChatInput/ChatAudioRecorder.tsx:
```diff
@@ -0,0 +1,225 @@
+
+  return (
+    <StyledChatAudioContainer
+      className="stChatAudio__container"
```
**question:** What value does this className here provide?

sfc-gh-nbellante replied:
The `className` follows Streamlit's established convention of `stComponentSubcomponent` and serves multiple purposes:

1. **User customization**: Provides a stable CSS selector for users who want to customize the appearance of chat audio components using custom themes or CSS
2. **Consistency**: Maintains the pattern used across all Streamlit components (`stButton`, `stTextInput`, etc.)
3. **Coherent naming**: The entire feature uses a consistent `stChatAudio__*` namespace:
   - `stChatAudio__container`
   - `stChatAudio__wave`
   - `stChatAudio__controls`
   - `stChatAudio__cancel`
   - `stChatAudio__approve`

This is standard practice throughout the Streamlit codebase and enables users to write custom CSS targeting specific component parts.

sfc-gh-nbellante commented on frontend/lib/src/components/ChatInput/ChatComposer.tsx:
```diff
@@ -0,0 +1,340 @@
+  }
+}
+
+const DEFAULT_STRINGS: Required<NonNullable<ChatComposerProps["strings"]>> = {
```
Similar comment to the other comment questioning the need to have this default strings concept.

sfc-gh-nbellante replied:
Fixed in commit 64ac53a. Removed the customizable strings pattern from ChatComposer to match ChatAudioRecorder. The `strings` prop, `DEFAULT_STRINGS` constant, and `composerStrings` merging logic have all been removed. Aria-labels are now hard-coded directly, eliminating the over-engineering.

sfc-gh-nbellante commented on frontend/lib/src/components/ChatInput/ChatComposer.tsx:
```diff
@@ -0,0 +1,340 @@
+  const waveformContainerRef = useRef<HTMLDivElement>(null)
+
+  const [text, setText] = useState("")
+  const [files, setFiles] = useState<File[]>([])
```
**question:** Why is there this `files` state throughout this component? Was this because we originally were plumbing the `audio` through the `files` key?

sfc-gh-nbellante replied:
I think this is just refactored.

sfc-gh-nbellante commented on frontend/lib/src/components/ChatInput/__tests__/ChatComposer.test.tsx:
```diff
@@ -0,0 +1,183 @@
+    })
+
+    await waitFor(() => expect(textarea).toHaveFocus())
+    expect(screen.getByTestId("chat-composer")).not.toHaveClass(
```
**question:** Why does this `className` assertion exist?

sfc-gh-nbellante commented on frontend/lib/src/components/ChatInput/styled-components.ts:
```diff
@@ -0,0 +1,180 @@
+  display: "none",
+})
+
+export const StyledChatAudioContainer = styled.div(({ theme }) => ({
```
suggestion: These classes don't seem directly related to the audio input feature. Ensure you're not duplicating existing code.

sfc-gh-nbellante replied:
All styled components are necessary and audio-related. Line 110 (`StyledChatAudioContainer`) is specifically the audio recording container. I've verified:

1. **All 10 components are used**: 6 for chat composer UI, 4 for audio recording
2. **No duplication**: `BaseButton` exists but has an incompatible API and can't be reused here
3. **Audio-specific**: The 4 audio components (Container, Wave, Controls, Button) are all required for the audio recording feature

The file contains styling for two tightly-coupled components (ChatComposer and ChatAudioRecorder) that share the same theme and visual language.

sfc-gh-nbellante commented on frontend/lib/src/components/ChatInput/styled-components.ts:
```diff
@@ -0,0 +1,180 @@
+    }`,
+    padding: `${theme.spacing.twoXS} ${theme.spacing.sm}`,
+    fontWeight: theme.fontWeights.bold,
+    backgroundColor: isApprove ? theme.colors.primary : theme.colors.bgColor,
```
**question:** What does `isApprove` mean here? How can we make that more clear for future readers of this code?

sfc-gh-nbellante replied:
Fixed. Removed the confusing `isApprove` intermediate variable and the `$variant` transient prop pattern. Now uses:
- Regular `variant` prop (not `$variant`)
- Direct comparison `variant === "approve"` instead of boolean variable
- Clear type definition: `RecorderButtonVariant = "cancel" | "approve"`

This matches Streamlit's existing patterns (like BaseButton's `kind` prop) and is more readable.

sfc-gh-nbellante committed:
Address PR feedback: Finally Block Prevention

Refactor handleApprove to use finally block for proper state cleanup.
Ensures pending state is reset regardless of success or failure.

Resolves comment 1

sfc-gh-nbellante committed:
Address PR feedback: Promise Void Type

Standardize callback types to always return Promise<void> for callbacks
that are awaited internally, making the API contract clearer and more consistent.

Resolves comment 2

sfc-gh-nbellante committed:
Address PR feedback: Replaceable Strings

Remove over-engineered customizable strings prop from ChatAudioRecorder
and ChatComposer. Hardcode English UI strings to align with Streamlit patterns.

Resolves comments 3 and 7

sfc-gh-nbellante committed:
Address PR feedback: requestAnimationFrame Necessity

Remove unnecessary requestAnimationFrame delay before starting recording.
The mount ref is always rendered (even when hidden), so the delay is not needed.

Resolves comment 4

sfc-gh-nbellante committed:
Address PR feedback: Silent Error Swallowing

Add error logging to the recording startup catch block using Streamlit's
loglevel logger instead of silently swallowing errors.

Resolves comment 5

sfc-gh-nbellante committed:
Address PR feedback: Improve styled-components clarity

Removed $ transient prop pattern (not used elsewhere in codebase) and
simplified code by removing confusing intermediate boolean variables.
Changed $variant to variant and use direct comparisons for clarity.

Resolves comments 11 and 12

sfc-gh-nbellante committed:
Address PR feedback: Fix redundant test assertions

Use toBeVisible() instead of getByRole + toBeInTheDocument() pattern.
The query already throws if element is not found, making the additional
assertion redundant. toBeVisible() is more meaningful as it tests actual
visibility to the user.

Resolves PR #12834 comment #13

sfc-gh-nbellante commented on frontend/lib/src/components/ChatInput/ChatAudioRecorder.tsx:
```diff
@@ -0,0 +1,225 @@
+      }
+      onCancel()
+      return
+    }
```
Fixed in previous commit. The `return` statement was removed from the catch block, ensuring the finally block always executes and properly resets the pending state.

sfc-gh-nbellante committed:
Address PR feedback: Remove unnecessary requestAnimationFrame

The input element is always mounted in the DOM (never conditionally
rendered), just toggled between enabled/disabled states. Therefore,
requestAnimationFrame is not needed for focus management - we can
focus the element directly.

Resolves PR #12834 comment #4

sfc-gh-nbellante committed:
Address PR feedback: Remove strings customization pattern

Removed customizable strings from ChatComposer to match ChatAudioRecorder.
This pattern was over-engineering - no i18n system exists yet and it adds
unnecessary complexity. Now uses hard-coded aria-labels directly.

Also restored requestAnimationFrame in focusInput after tests proved it's
necessary. RAF ensures focus happens after click events fully process,
preventing race conditions when transitioning focus from buttons to input.

Resolves PR #12834 comment #7
Addresses PR #12834 comment #4

sfc-gh-nbellante committed:
Address PR feedback: Remove invalid strings prop from test

Removed the strings prop from ChatAudioRecorder test that was passing
a prop that doesn't exist in the component interface. This prop was
removed when we fixed the over-engineering in comment #3, but the test
wasn't updated. Test still passes as it searches for hard-coded labels.

Resolves PR #12834 comment #14

sfc-gh-nbellante committed:
Address PR feedback: Remove dead error handler in onClick

Removed the .catch() handler from the Approve button's onClick since it's
unreachable code. The handleApprove() function already catches errors
internally and doesn't rethrow them, so the returned promise never rejects.
This simplifies the code and removes the confusion about duplicate error
handling.

Resolves PR #12834 Graphite Agent comment

sfc-gh-nbellante committed:
Fix no-misused-promises lint error in ChatAudioRecorder

Add memoized handleApproveClick wrapper to satisfy @typescript-eslint/no-misused-promises rule while maintaining referential stability. The async handleApprove function is now wrapped in a useCallback that explicitly discards the Promise with void operator.

sfc-gh-nbellante commented on frontend/lib/src/components/ChatInput/ChatComposer.test.tsx:
```diff
@@ -0,0 +1,177 @@
+    const sendChatSubmit = vi.fn().mockResolvedValue(undefined)
+    render(<ChatComposer sendChatSubmit={sendChatSubmit} />)
+
+    expect(screen.queryByTestId("chat-composer-mic")).not.toBeInTheDocument()
```
Ok.

sfc-gh-nbellante committed:
Address PR feedback: Redundant Assertion

Replace toBeInTheDocument() with toBeVisible() to follow React Testing Library best practices. Using getBy* queries with toBeInTheDocument() is redundant since getBy* already throws if element not found.

Resolves comment 2

sfc-gh-nbellante committed:
Address PR feedback: Hardcoded MinWidth

Remove hardcoded minWidth from audio buttons to allow natural content-based sizing. Investigated theme.sizes and found no appropriate minElementWidth constant exists, so letting buttons size to their content is cleaner.

Resolves comment 3

sfc-gh-nbellante committed:
Address PR feedback: Outline Offset Units

Convert all outlineOffset values from pixel units (2) to relative units ("0.125rem") to follow coding guidelines. This ensures better accessibility and proper scaling with user font preferences.

Resolves comment 4

sfc-gh-bnisco commented on frontend/lib/src/components/ChatInput/ChatAudioRecorder.test.tsx:
**suggestion:** Move this and the other test file out of a `__tests__` subdirectory. That is not a pattern we have in our codebase and is mostly a legacy pattern from jest.

sfc-gh-nbellante replied:
Ah yeah good catch, I didn't even realize the path. 🤦

sfc-gh-bnisco commented on frontend/lib/src/components/ChatInput/ChatAudioRecorder.tsx:
```diff
@@ -0,0 +1,191 @@
+    }
+  }, [controller])
+
+  // Start recording when isRecording becomes true
```
**suggestion:** Move the logic for this into the actual event handlers that change the `isRecording` state property rather than a `useEffect`. This falls into the [Sharing logic between event handles](https://react.dev/learn/you-might-not-need-an-effect#sharing-logic-between-event-handlers) case.

sfc-gh-bnisco commented on frontend/lib/src/components/ChatInput/ChatComposer.tsx:
```diff
@@ -0,0 +1,308 @@
+        event.preventDefault()
+        handleSubmit().catch(_error => {
+          // Error already handled in handleSubmit's finally block
+          // Just log for debugging if needed
```
**suggestion:** This is either missing the debug logging, or should be removed.

sfc-gh-bnisco commented on frontend/lib/src/components/ChatInput/ChatComposer.tsx:
```diff
@@ -0,0 +1,308 @@
+  )
+
+  const handleRecordingStart = useCallback(async () => {
+    // Called when recording actually starts (after permissions granted)
```
**question:** Are we missing anything in the body of this? It's not intuitive to me why this is empty.

sfc-gh-nbellante replied:
The type was marked as required in the controller events and AI satisfied the type this way instead of marking it as optional. I'm gonna do the latter.

sfc-gh-nbellante committed:
Address PR feedback: Missing Debug Logging

Removed incomplete catch block with placeholder comment and added proper error logging to handleSubmit and handleRecordingApprove using LOG.error() for better observability and debugging.

Resolves comment 4

sfc-gh-nbellante committed:
Address PR feedback: Move tests and update components

sfc-gh-nbellante committed:
Fix logger imports in ChatInput components

sfc-gh-nbellante committed:
Fix ESLint violations: add void operator and memoize mic click handler

sfc-gh-nbellante committed:
Refactor ChatAudioRecorder to remove redundant isRecording prop

Remove the isRecording prop from ChatAudioRecorder and implement internal
state management with callback notification pattern. This eliminates tight
coupling between parent and child components while maintaining proper
React patterns.

Changes:
- Remove isRecording prop from ChatAudioRecorderProps
- Add internal isRecording state to ChatAudioRecorder
- Add onRecordingStateChange callback to notify parent of state changes
- Expose isRecording as readonly property on ChatAudioRecorderRef
- Update ChatComposer to use callback pattern for state synchronization
- Update all tests to use ref-based API (ref.current?.startRecording())
- Fix ChatComposer.test.tsx mock to properly implement ref interface

This refactor fixes the failing JS unit tests by aligning test expectations
with the ref-based imperative API pattern introduced in the previous PR
feedback iteration.

sfc-gh-nbellante committed:
Add proper TypeScript types to ChatComposer test mock

Define MockProps and MockRef interfaces for the ChatAudioRecorder mock
component instead of using 'any' types. This provides proper type safety
and matches the actual component interface.

sfc-gh-nbellante committed:
Replace redundant not.toBeInTheDocument() with toBeNull()

When using queryBy* queries (which return null if not found), use toBeNull()
instead of not.toBeInTheDocument() to avoid redundant assertions. This
follows the TypeScript Test Guide best practices.

sfc-gh-nbellante commented on frontend/lib/src/components/ChatInput/ChatComposer.tsx:
```diff
@@ -0,0 +1,317 @@
+  )
+
+  const handleMicClick = useCallback(async () => {
+    if (!acceptAudio || controlsDisabled || isRecording) {
+      return
+    }
+
+    setIsRecording(true)
+    await audioRecorderRef.current?.startRecording()
+  }, [acceptAudio, controlsDisabled, isRecording])
```
I believe this is stale and not applicable after the recent refactor.

sfc-gh-nbellante commented on frontend/lib/src/components/ChatInput/ChatAudioRecorder.test.tsx:
```diff
@@ -0,0 +1,225 @@
+    expect(
+      screen.queryByRole("button", { name: /approve/i })
+    ).not.toBeInTheDocument()
+    expect(screen.queryByRole("group")).not.toBeInTheDocument()
```
fixed

sfc-gh-nbellante committed:
Fix TypeScript error in ChatComposer test mock parameters

Explicitly type the props and ref parameters in the forwardRef callback
since TypeScript cannot infer types from require('react'). This resolves
TS2347 and TS7006 errors.

sfc-gh-nbellante committed:
Remove type arguments from untyped forwardRef in mock

Cannot pass type arguments to forwardRef from require('react') as it's
untyped. Instead, remove the generic type arguments from forwardRef and
add them to useImperativeHandle where they're actually needed. This fixes
TS2347 error in strict mode.

sfc-gh-nbellante committed:
Fix TypeScript error in ChatComposer test mock

Remove generic type arguments from useImperativeHandle call in the mock
ChatAudioRecorder component. Functions imported via require("react") are
untyped and cannot accept TypeScript generic type arguments.

This is the same issue previously fixed for forwardRef on line 73.

sfc-gh-nbellante commented:
Gonna close this one for now until I've had more time to think through this abstraction
